rfcs/text/0010_service_status.md

```diff
@@ -0,0 +1,304 @@
+  http: ServiceStatus;
+  savedObjects: ServiceStatus;
+  uiSettings: ServiceStatus;
+  metrics: ServiceStatus;
+}
+```
+
+## API Design
+
+### Plugin API
+
+```ts
+/**
+ * The API exposed to plugins on CoreSetup.status
+ */
+interface StatusSetup {
+  /**
+   * Allows a plugin to specify a custom status dependent on its own criteria.
+   * Completely overrides the default inherited status.
+   */
+  set(status$: Observable<ServiceStatus>): Observable<ServiceStatus>;
+
+  /**
+   * Current status for all Core services.
+   */
+  core$: Observable<CoreStatus>;
+
+  /**
+   * Current status for all dependencies of the current plugin.
+   * Each key of the `Record` is a plugin id.
+   */
+  plugins$: Observable<Record<string, ServiceStatus>>;
+
+  /**
+   * The default status of the plugin as specified in the "Status inheritance"
+   * section below.
+   * 
+   * Useful when overriding the defaults with `set`.
+   */
+  inherited$: Observable<ServiceStatus>;
+}
+```
+
+### HTTP API
+
+The HTTP endpoint should return basic information about the Kibana node as well as the overall system status and the status of each individual system.
+
+This API does not need to include UI-specific details like the existing API such as `uiColor` and `icon`.
+
+```ts
+/**
+ * Response type for the endpoint: GET /api/status
+ */
+interface StatusResponse {
+  /** server.name */
+  name: string;
+  /** server.uuid */
+  uuid: string;
+  /** Currently exposed by existing status API */
+  version: {
+    number: string;
+    build_hash: string;
+    build_number: number;
+    build_snapshot: boolean;
+  };
+  /** Similar format to existing API, but slightly different shape */
+  status: {
+    overall: ServiceStatus;
+    core: Record<string, ServiceStatus>;
+    plugins: Record<string, ServiceStatus>;
+  }
+}
+```
+
+## Behaviors
+
+### Levels
+
+Each member of the `ServiceStatusLevel` enum has specific behaviors associated with it:
+- **`available`**:
+  - All endpoints and apps associated with the service are accessible
+- **`degraded`**:
+  - All endpoints and apps are available by default
+  - Some APIs may return `503 Unavailable` responses. This must be implemented directly by the service.
+- **`unavailable`**:
+  - All endpoints (with some exceptions in Core) in Kibana return a `503 Unavailable` responses by default. This is automatic.
+  - When trying to access any app associated with the unavailable service, the user is presented with an error UI with detail about the outage.
+- **`fatal`**:
+  - All endpoints (with some exceptions in Core) in Kibana return a `503 Unavailable` response by default. This is automatic.
+  - All applications redirect to the system-wide status page with detail about which services are down and any relevant detail.
+
+### Status inheritance
+
+By default, plugins inherit their status from all Core services and their dependencies on other plugins. The default status for a plugin follows this algorithm:
+```js
+// pseudo-code
```

Comment:
optional: using a table representation would help to understand the logic at first glance

Reply:
Added a table alongside the pseudocode